sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala

```diff
@@ -107,6 +109,25 @@ object HiveThriftServer2 extends Logging {
       } else {
         None
       }
+
+      // HiveThriftServer2 HA
+      val metaVersion = IsolatedClientLoader.hiveVersion(
+          SparkSQLEnv.sqlContext.conf.getConf(HiveUtils.HIVE_METASTORE_VERSION))
+      metaVersion match {
+        case hive.v12 | hive.v13 =>
+          logWarning("HiveThriftServer2 HA is compatible with Hive metastore version is " +
+              "0.14.0 or higher.")
+        case _ =>
+          if (executionHive.conf.getBoolVar(
+            ConfVars.HIVE_SERVER2_SUPPORT_DYNAMIC_SERVICE_DISCOVERY)) {
+            val method = server.getClass.getSuperclass.getDeclaredMethod(
+                "addServerInstanceToZooKeeper",
```

Comment:
The ```addServerInstanceToZooKeeper``` method has been removed in SPARK-15004. 
How can this work?
You can take a look at the discussion of SPARK-11100 first.

Reply:
Because HiveThriftServer2 not support HA based on Hive, so SPARK-11100 occurs.
I have investigated the `addServerInstanceToZooKeeper` method in Hive，it was supported from version 0.14.0 to version 3.1.1

Reply:
HiveThriftServer2 inherits HiveServer2, and HiveServer2 does not currently have the addServerInstanceToZooKeeper method.
Call getDeclaredMethod("addServerInstanceToZooKeeper") directly, will not throw NoSuchMethodException now?

Reply:
OK. You said is correct based on HiveServer2 that contains by `hive-thriftserver` module.
There exists issue as follows:
First，why the `hive-thriftserver` module copy the code about `HiveServer2` from https://github.com/apache/hive?
Second, every company uses different Hive version. For example，my company uses Hive contains 1.2.1 and 2.X.So we first uses our Hive library instead of `hive-thriftserver` module carrying.
Third,SPARK-15004 looks arbitrary and not enough discussed.